rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py

```diff
@@ -0,0 +1,105 @@
+        where the user sends and receives any asset. The spend asset is the swap executed and
+        the received is the result of calling the swap method in the uniswap contract.
+        We need to make sure that the events related to the swap are consecutive and for that
+        we make use of the maybe_reshuffle_events.
+        """
+        if tx_log.topics[0] == SWAP_SIGNATURE:
+            # When the router chains multiple swaps in one transaction only the last swap has
+            # the buyer in the topic. In that case we know it is the last swap and the receiver is
+            # the user
+            maybe_buyer = hex_or_bytes_to_address(tx_log.topics[2])
+            out_event = in_event = None
+
+            # amount_in is the amount that enters the pool and amount_out the one
+            # that leaves the pool
+            amount_in_0, amount_in_1 = hex_or_bytes_to_int(tx_log.data[0:32]), hex_or_bytes_to_int(tx_log.data[32:64])  # noqa: E501
+            amount_out_0, amount_out_1 = hex_or_bytes_to_int(tx_log.data[64:96]), hex_or_bytes_to_int(tx_log.data[96:128])  # noqa: E501
+            amount_in, amount_out = amount_in_0, amount_out_0
+            if amount_in == ZERO:
+                amount_in = amount_in_1
+            if amount_out == ZERO:
+                amount_out = amount_out_1
+
+            received_eth = ZERO
+            for event in decoded_events:
+                if event.asset == A_ETH and event.event_type == HistoryEventType.RECEIVE:
+                    received_eth += event.balance.amount
+
+            for event in decoded_events:
+                # When we look for the spend event we have to take into consideration the case
+                # where not all the ETH is converted. The ETH that is not converted is returned
+                # in an internal transaction to the user.
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    (
+                        event.balance.amount == (spend_eth := asset_normalized_value(amount_in, event.asset)) or  # noqa: E501
+                        event.asset == A_ETH and spend_eth + received_eth == event.balance.amount
+                    )
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.SPEND
+                    event.counterparty = CPT_UNISWAP_V2
+                    event.notes = f'Swap {event.balance.amount} {event.asset.symbol} in uniswap-v2 from {event.location_label}'  # noqa: E501
+                    out_event = event
+                elif (
+                    (maybe_buyer == transaction.from_address or event.asset == A_ETH) and
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.balance.amount == asset_normalized_value(amount_out, event.asset)
+                ):
+                    event.event_type = HistoryEventType.TRADE
```

Comment:
Won't it be better to maybe have one trade instead of two? (If I understand correctly, in the current implementation there is going to be one trade for spend and one for receive for each swap, right?)

Also, maybe will make sense to even combine all the swaps into a single one? So if there are multiple underlying swaps, they jus turn into a single one?

Reply:
This is how all swaps are. It's not a trade. All decoded transactions boil down to single base events. This is how our decoders work and how all base rotki events should be at the end.

Each event is an asset going in or an asset going out. Some events. like here, get comined together in swaps, or trades.